include/swift-c/DependencyScan/DependencyScan.h

```diff
@@ -492,6 +539,130 @@ SWIFTSCAN_PUBLIC swiftscan_string_ref_t
+SWIFTSCAN_PUBLIC bool
+swiftscan_cache_compilation_has_output_kind(swiftscan_cached_compilation_t,
+                                            swiftscan_output_kind_t,
+                                            swiftscan_string_ref_t *error);
+
+/// Make the cache compilation available globally. \c callback will be called
+/// on completion.
+/// \c swiftscan_cache_cancellation_token_t will return an token that can be
+/// used to cancel the async operation. The token needs to be freed by caller
+/// using `swiftscan_cache_cancellation_token_dispose`.
+SWIFTSCAN_PUBLIC void swiftscan_cache_make_global_async(
+    swiftscan_cas_t cas, const char *key, void *ctx,
+    void (*callback)(void *ctx, swiftscan_string_ref_t error),
+    swiftscan_cache_cancellation_token_t *);
+
+/// Cancel the async cache action that is associated with token.
+SWIFTSCAN_PUBLIC void
+    swiftscan_cache_action_cancel(swiftscan_cache_cancellation_token_t);
+
+/// Dispose the cancellation token.
+SWIFTSCAN_PUBLIC void swiftscan_cache_cancellation_token_dispose(
+    swiftscan_cache_cancellation_token_t);
+
+/// Create a swift cached compilation replay instance with its command-line
+/// invocation. Return nullptr when errors occurs and the error message is
+/// returned via \c error parameter and its caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_replay_instance_t
+swiftscan_cache_replay_instance_create(int argc, const char **argv,
+                                       swiftscan_string_ref_t *error);
+
+/// Dispose swift cached compilation replay instance.
+SWIFTSCAN_PUBLIC void
+    swiftscan_cache_replay_instance_dispose(swiftscan_cache_replay_instance_t);
+
+/// Replay the cached compilation using cached compliation replay instance.
+/// If the return value is SWIFTSCAN_CACHE_RESULT_ERROR, the error message is
+/// returned via \c error parameter and its caller needs to free the message
+/// using `swiftscan_string_dispose`.
+SWIFTSCAN_PUBLIC swiftscan_cache_result_t swiftscan_cache_replay_compilation(
+    swiftscan_cas_t cas, swiftscan_cache_replay_instance_t,
+    swiftscan_cached_compilation_t, swiftscan_string_ref_t *error);
+
+/// Async replay the cached compilation. Callback function will be provided
+/// by the cache replay result, and its stdout, stderr, and any error message.
+/// \c ctx is an opaque value that passed to the callback and
+/// \c swiftscan_cache_cancellation_token_t will return an token that can be
+/// used to cancel the async operation. The token needs to be freed by caller
+/// using `swiftscan_cache_cancellation_token_dispose`.
+SWIFTSCAN_PUBLIC void swiftscan_cache_replay_compilation_async(
+    swiftscan_cas_t cas, swiftscan_cache_replay_instance_t,
+    swiftscan_cached_compilation_t, void *ctx,
+    void (*callback)(void *ctx, swiftscan_cache_result_t,
+                     swiftscan_string_ref_t std_out,
+                     swiftscan_string_ref_t std_err,
```

Comment:
The synchronous version (`swiftscan_cache_replay_compilation`) is not providing `std_out` and `std_err`.
I'd suggest to create an opaque object `swiftscan_cache_replay_result` that could encapsulate these kind of data.